### DISCUSSION TITLE
SetBody behaves differently depending on the method used

### DISCUSSION THREAD

**Discussion by wsw70** [go-resty/resty#581] [Q&A] [answered]
I am using `resty`  to send a POST to a home service (Home Assistant, but it does no matter).

The version that works (explicit map):

```go
package main

import (
	"fmt"

	"github.com/go-resty/resty/v2"
)

func sendToHass(sensor string, value string, attributes map[string]string) error {
	resp, err := resty.New().R().
		SetAuthToken("eyXXTU").
		SetBody(map[string]interface{}{
		 	"state":      "hello",
		 	"attributes": map[string]string{},
	        }).
		Post(fmt.Sprintf("https://hass.example.com/api/states/sensor.dash_%v", sensor))
	if (resp.StatusCode() == 200 || resp.StatusCode() == 201) && err == nil {
		return nil
	} else {
		return fmt.Errorf("unexpected answer (not 200 or 201): %v (err: %v)", resp.StatusCode(), err)
	}
}
```

I then wanted to replace the `map` with a better looking type:

```go
package main

import (
	"fmt"

	"github.com/go-resty/resty/v2"
)

type State struct {
	State string
	// Attributes map[string]string
}

func sendToHass(sensor string, value string, attributes map[string]string) error {
	resp, err := resty.New().R().
		SetAuthToken("eyXXXXXU").
		SetBody(State{
			State: "hello",
			// Attributes: attributes,
		}).
		Post(fmt.Sprintf("https://hass.example.com/api/states/sensor.dash_%v", sensor))
	if (resp.StatusCode() == 200 || resp.StatusCode() == 201) && err == nil {
		return nil
	} else {
		return fmt.Errorf("unexpected answer (not 200 or 201): %v (err: %v)", resp.StatusCode(), err)
	}
}
```

This returns a `400` (it does not matter whether `Attributes` is commented out or not):

    {"level":"error","time":"2022-09-20T12:16:32+02:00","message":"cannot send outside temperature to HA: unexpected answer (not 200 or 201): 400 (err: <nil>)"}

Since this is the only change in the code there must be something strange with the body, which should be sent out as

```json
{
  "state": "something",
  "attributes": {}
}
```

(with or without `"attributes"`)

I would appreciate very much any pointers.

**Comment by kecci** [comment, marked as answer]
Hi @wsw70,

Disclaimer: this answer reference to my project experience before.

I've try both of your code to using [webhook.site](https://webhook.site).

🔭  ### Firstly, i check the using `setBody(map[string]interface{})`, and the result was:

<img width="1512" alt="Screen Shot 2022-10-07 at 03 32 31" src="https://user-images.githubusercontent.com/25241073/194412945-5ba1d105-c738-4f2d-b22e-7f7237235dcb.png">

The body sent was: 
```json
{
  "attributes": {},
  "state": "hello"
}
```

🔭 ### Secondly, i check the using `setBody(State{})`, and the result was:

<img width="1512" alt="Screen Shot 2022-10-07 at 03 31 24" src="https://user-images.githubusercontent.com/25241073/194413679-f415f809-f994-4eac-9540-8acdabb78740.png">

The body sent was: 
```json
{
  "State": "hello"
}
```

🚀  ### Conclusion, if you want using struct to body json, use the `struct tags` and convert the struct to json.

Fix the code into like this:
```go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/go-resty/resty/v2"
)

// You have to add the struct tags on here
type State struct {
	State      string            `json:"state"`
	Attributes map[string]string `json:"attributes"`
}

func sendToHass(sensor string, value string, attributes map[string]string) error {
	// Prepare your struct
	state := State{
		State:      "hello",
		Attributes: attributes,
	}

	// Convert struct to json byte
	stateByte, _ := json.Marshal(state)

	resp, err := resty.New().R().
		SetAuthToken("eyXXXXXU").
		SetBody(string(stateByte)). // Use in here. Also mapping type of byte to string
		Post("https://webhook.site/9e0d8a64-0ed1-4e28-ae32-ddd26fae5e1d")
	if (resp.StatusCode() == 200 || resp.StatusCode() == 201) && err == nil {
		return nil
	} else {
		return fmt.Errorf("unexpected answer (not 200 or 201): %v (err: %v)", resp.StatusCode(), err)
	}
}

func main() {
	sendToHass("sensor", "value", map[string]string{})
}

```

You can see more about struct tags on here: https://www.digitalocean.com/community/tutorials/how-to-use-struct-tags-in-go

I hope this will help your discussions.

**Reply by wsw70** [reply]
Thank you very much, I completely forgot that the case would be different. Thanks!

**Reply by kecci** [reply]
Your welcome @wsw70 .
